### PR TITLE
[WIP] Refactor

### LINKED ISSUE #15: Refactoring
Hi, I  am trying to make a Dockerfile and refactor directory structures.

In my branch [docker](https://github.com/alex-ht/CAT/tree/docker), I have finished a [dockerfile](https://github.com/alex-ht/CAT/blob/docker/docker/Dockerfile) ~~based on kaldi's official `gpu-latest`  image. The pytorch is at 1.2.0 version, because the base image, kaldiasr/kaldi:gpu-latest, is staying at CUDA10.0. I'm not familiar with  pytorch, so I just copy the instrunction from official site [here](https://pytorch.org/get-started/previous-versions/).~~

Update:  I chose pytorch 1.5 as base image now. 

After those works, I found out that the scripts in wsj/steps and utils are not following kaldi's guidline.
Some of them were copied from old version of kaldi, Eesen, etc. I would suggest untracking typeical scripts from old Kaldi and moving remains into [scripts/ctc-crf](https://github.com/alex-ht/CAT/tree/docker/scripts/ctc-crf). Every egs has two soft links, `steps` and `utils`, to kaldi's wsj, and one soft link goes to `scripts/ctc-crf `.

Do you agree with above changes? I'm happy to make PR.

### PR DESCRIPTION
**Pull request by alex-ht**
Hi, I have finished most of refactor jobs. It should be able to run with pytorch 1.5.

Here is task list:

- [x] move all shared scripts to `scripts/ctc-crf`
- [x] Added option '--arch' to select network architecture
- [x] `calculate_logits.py` is called by `decode.sh`
- [x] added Dockerfile
- [ ] refine `egs`
  - [x] librispeech
  - [x] aishell
  - [x] WSJ (needs help)
  - [ ] switchboard (needs help)
- [x] added `egs/formosa` (alex)
- [x] refine `decode.sh`: load model configuration.(alex)

Resolve #15

### CODE REVIEW

**Comment by alex-ht** [comment]
hi @aky15 , you can edit this branch now.

**Comment by aky15** [comment]
thank you Alex @alex-ht. I'll review it soon.

**Comment by aky15** [comment]
Thank you Alex @alex-ht .  The structure is much more clear and some modification (e.g. the --arch option ) is quite necessary and helpful. So far, I've found some small issues.
1. It seems there is no definition of the option "--calculate-logits-opts" in scripts/ctc-crf/decode.sh.
2. if ctc-crf/decode.sh is shared by all egs, the decode.sh in dictionary local for all egs should be deleted. (make sure ctc-crf/decode.sh is of universal applicability and transplantability.)
3. the argument "dir" can be replaced with "--model_path" to stay consistent with other arguments in ctc-crf/train.py.
4. Have you get the results for egs/formosa? I suggest to use a RESULT file to record your experimental results and I'm very interested in it.

**Comment by alex-ht** [comment]
> Thank you Alex @alex-ht . The structure is much more clear and some modification (e.g. the --arch option ) is quite necessary and helpful. So far, I've found some small issues.
> 
>     1. It seems there is no definition of the option "--calculate-logits-opts" in scripts/ctc-crf/decode.sh.
>     2. if ctc-crf/decode.sh is shared by all egs, the decode.sh in dictionary local for all egs should be deleted. (make sure ctc-crf/decode.sh is of universal applicability and transplantability.)
I will rewrite [here](https://github.com/alex-ht/CAT/blob/6370bfcdb5100aecf46961988dbf913f9e77dd76/scripts/ctc-crf/train.py#L94-L103) into bash compatible style, then remove option "--calculate-logits-opts".
>     3. the argument "dir" can be replaced with "--model_path" to stay consistent with other arguments in ctc-crf/train.py.
`dir` is not optional, so it should be a positional argument.(?)
>     4. Have you get the results for egs/formosa? I suggest to use a RESULT file to record your experimental results and I'm very interested in it.
RESULT was copied from kaldi. I'm still tuning hyperparameters. 😅

**Comment by aky15** [comment]
Alex @alex-ht 你好，代码我已经审阅过了，可以合并到master分支里。关于decode.sh，我建议可以把inference的代码（即calculate_logits.py）拿到顶层脚本run.sh里。因为在一些egs里（例如WSJ），inference得到的结果会被不同解码图（decode FST）共用，如果calculate_logits.py放到decode.sh里，会造成重复计算。当然也可以设计一些机制来避免重复inference ：）

**Comment by alex-ht** [comment]
谢谢！最近公事繁忙，会晚一点处理。

**Comment by alex-ht** [comment]
hi, @aky15 
可以请你看一下WSJ的范例吗？我用控制stage参数来避免重复计算。
另外score.sh在acwt参数的设计，我改成kaldi的方法。产生lattice时让acwt放大，score.sh就可以沿用整数的lmwt。

**Comment by alex-ht** [comment]
Thanks!
